File: frontend/src/components/benchmark/content/red_black_tree/implementations/red_black_tree.cs
using System;
using System.Collections.Generic;

public class RedBlackTree
{
    private class Node
    {
        public int Key;
        public bool Color; // true = RED, false = BLACK
        public Node Left;
        public Node Right;
        public Node Parent;

        public Node(int key, bool color = true)
        {
            Key = key;
            Color = color;
            Left = null;
            Right = null;
            Parent = null;
        }
    }

    private Node root;
    private Node NIL;

    public RedBlackTree()
    {
        NIL = new Node(0, false);
        root = NIL;
    }

    private void LeftRotate(Node x)
    {
        Node y = x.Right;
        x.Right = y.Left;
        if (y.Left != NIL)
            y.Left.Parent = x;
        y.Parent = x.Parent;
        if (x.Parent == null)
            root = y;
        else if (x == x.Parent.Left)
            x.Parent.Left = y;
        else
            x.Parent.Right = y;
        y.Left = x;
        x.Parent = y;
    }

    private void RightRotate(Node x)
    {
        Node y = x.Left;
        x.Left = y.Right;
        if (y.Right != NIL)
            y.Right.Parent = x;
        y.Parent = x.Parent;
        if (x.Parent == null)
            root = y;
        else if (x == x.Parent.Right)
            x.Parent.Right = y;
        else
            x.Parent.Left = y;
        y.Right = x;
        x.Parent = y;
    }

    private void InsertFixup(Node z)
    {
        while (z.Parent != null && z.Parent.Color == true)
        {
            if (z.Parent == z.Parent.Parent.Left)
            {
                Node y = z.Parent.Parent.Right;
                if (y.Color == true)
                {
                    z.Parent.Color = false;
                    y.Color = false;
                    z.Parent.Parent.Color = true;
                    z = z.Parent.Parent;
                }
                else
                {
                    if (z == z.Parent.Right)
                    {
                        z = z.Parent;
                        LeftRotate(z);
                    }
                    z.Parent.Color = false;
                    z.Parent.Parent.Color = true;
                    RightRotate(z.Parent.Parent);
                }
            }
            else
            {
                Node y = z.Parent.Parent.Left;
                if (y.Color == true)
                {
                    z.Parent.Color = false;
                    y.Color = false;
                    z.Parent.Parent.Color = true;
                    z = z.Parent.Parent;
                }
                else
                {
                    if (z == z.Parent.Left)
                    {
                        z = z.Parent;
                        RightRotate(z);
                    }
                    z.Parent.Color = false;
                    z.Parent.Parent.Color = true;
                    LeftRotate(z.Parent.Parent);
                }
            }
        }
        root.Color = false;
    }

    private void Transplant(Node u, Node v)
    {
        if (u.Parent == null)
            root = v;
        else if (u == u.Parent.Left)
            u.Parent.Left = v;
        else
            u.Parent.Right = v;
        v.Parent = u.Parent;
    }

    private Node Minimum(Node node)
    {
        while (node.Left != NIL)
            node = node.Left;
        return node;
    }

    private void DeleteFixup(Node x)
    {
        while (x != root && x.Color == false)
        {
            if (x == x.Parent.Left)
            {
                Node w = x.Parent.Right;
                if (w.Color == true)
                {
                    w.Color = false;
                    x.Parent.Color = true;
                    LeftRotate(x.Parent);
                    w = x.Parent.Right;
                }
                if (w.Left.Color == false && w.Right.Color == false)
                {
                    w.Color = true;
                    x = x.Parent;
                }
                else
                {
                    if (w.Right.Color == false)
                    {
                        w.Left.Color = false;
                        w.Color = true;
                        RightRotate(w);
                        w = x.Parent.Right;
                    }
                    w.Color = x.Parent.Color;
                    x.Parent.Color = false;
                    w.Right.Color = false;
                    LeftRotate(x.Parent);
                    x = root;
                }
            }
            else
            {
                Node w = x.Parent.Left;
                if (w.Color == true)
                {
                    w.Color = false;
                    x.Parent.Color = true;
                    RightRotate(x.Parent);
                    w = x.Parent.Left;
                }
                if (w.Right.Color == false && w.Left.Color == false)
                {
                    w.Color = true;
                    x = x.Parent;
                }
                else
                {
                    if (w.Left.Color == false)
                    {
                        w.Right.Color = false;
                        w.Color = true;
                        LeftRotate(w);
                        w = x.Parent.Left;
                    }
                    w.Color = x.Parent.Color;
                    x.Parent.Color = false;
                    w.Left.Color = false;
                    RightRotate(x.Parent);
                    x = root;
                }
            }
        }
        x.Color = false;
    }

    private Node Search(Node node, int key)
    {
        if (node == NIL || key == node.Key)
            return node;
        if (key < node.Key)
            return Search(node.Left, key);
        return Search(node.Right, key);
    }

    private void Inorder(Node node, List<int> result)
    {
        if (node != NIL)
        {
            Inorder(node.Left, result);
            result.Add(node.Key);
            Inorder(node.Right, result);
        }
    }

    public void Insert(int key)
    {
        Node z = new Node(key);
        z.Left = NIL;
        z.Right = NIL;

        Node y = null;
        Node x = root;

        while (x != NIL)
        {
            y = x;
            if (z.Key < x.Key)
                x = x.Left;
            else
                x = x.Right;
        }

        z.Parent = y;
        if (y == null)
            root = z;
        else if (z.Key < y.Key)
            y.Left = z;
        else
            y.Right = z;

        InsertFixup(z);
    }

    public void Delete(int key)
    {
        Node z = Search(root, key);
        if (z == NIL) return;

        Node y = z;
        bool yOriginalColor = y.Color;
        Node x;

        if (z.Left == NIL)
        {
            x = z.Right;
            Transplant(z, z.Right);
        }
        else if (z.Right == NIL)
        {
            x = z.Left;
            Transplant(z, z.Left);
        }
        else
        {
            y = Minimum(z.Right);
            yOriginalColor = y.Color;
            x = y.Right;
            if (y.Parent == z)
            {
                x.Parent = y;
            }
            else
            {
                Transplant(y, y.Right);
                y.Right = z.Right;
                y.Right.Parent = y;
            }
            Transplant(z, y);
            y.Left = z.Left;
            y.Left.Parent = y;
            y.Color = z.Color;
        }

        if (yOriginalColor == false)
            DeleteFixup(x);
    }

    public bool Search(int key)
    {
        return Search(root, key) != NIL;
    }

    public List<int> Inorder()
    {
        List<int> result = new List<int>();
        Inorder(root, result);
        return result;
    }
}

File: frontend/src/components/benchmark/content/red_black_tree/implementations/red_black_tree.php
<?php

class RBNode {
    public int $key;
    public string $color;
    public ?RBNode $left;
    public ?RBNode $right;
    public ?RBNode $parent;

    public function __construct(int $key, string $color = 'RED') {
        $this->key = $key;
        $this->color = $color;
        $this->left = null;
        $this->right = null;
        $this->parent = null;
    }
}

class RedBlackTree {
    private ?RBNode $root;
    private RBNode $NIL;

    public function __construct() {
        $this->NIL = new RBNode(0, 'BLACK');
        $this->root = $this->NIL;
    }

    private function leftRotate(RBNode $x): void {
        $y = $x->right;
        $x->right = $y->left;
        if ($y->left !== $this->NIL) {
            $y->left->parent = $x;
        }
        $y->parent = $x->parent;
        if ($x->parent === null) {
            $this->root = $y;
        } else if ($x === $x->parent->left) {
            $x->parent->left = $y;
        } else {
            $x->parent->right = $y;
        }
        $y->left = $x;
        $x->parent = $y;
    }

    private function rightRotate(RBNode $x): void {
        $y = $x->left;
        $x->left = $y->right;
        if ($y->right !== $this->NIL) {
            $y->right->parent = $x;
        }
        $y->parent = $x->parent;
        if ($x->parent === null) {
            $this->root = $y;
        } else if ($x === $x->parent->right) {
            $x->parent->right = $y;
        } else {
            $x->parent->left = $y;
        }
        $y->right = $x;
        $x->parent = $y;
    }

    private function insertFixup(RBNode $z): void {
        while ($z->parent !== null && $z->parent->color === 'RED') {
            if ($z->parent === $z->parent->parent->left) {
                $y = $z->parent->parent->right;
                if ($y->color === 'RED') {
                    $z->parent->color = 'BLACK';
                    $y->color = 'BLACK';
                    $z->parent->parent->color = 'RED';
                    $z = $z->parent->parent;
                } else {
                    if ($z === $z->parent->right) {
                        $z = $z->parent;
                        $this->leftRotate($z);
                    }
                    $z->parent->color = 'BLACK';
                    $z->parent->parent->color = 'RED';
                    $this->rightRotate($z->parent->parent);
                }
            } else {
                $y = $z->parent->parent->left;
                if ($y->color === 'RED') {
                    $z->parent->color = 'BLACK';
                    $y->color = 'BLACK';
                    $z->parent->parent->color = 'RED';
                    $z = $z->parent->parent;
                } else {
                    if ($z === $z->parent->left) {
                        $z = $z->parent;
                        $this->rightRotate($z);
                    }
                    $z->parent->color = 'BLACK';
                    $z->parent->parent->color = 'RED';
                    $this->leftRotate($z->parent->parent);
                }
            }
        }
        $this->root->color = 'BLACK';
    }

    private function transplant(RBNode $u, RBNode $v): void {
        if ($u->parent === null) {
            $this->root = $v;
        } else if ($u === $u->parent->left) {
            $u->parent->left = $v;
        } else {
            $u->parent->right = $v;
        }
        $v->parent = $u->parent;
    }

    private function minimum(RBNode $node): RBNode {
        while ($node->left !== $this->NIL) {
            $node = $node->left;
        }
        return $node;
    }

    private function deleteFixup(RBNode $x): void {
        while ($x !== $this->root && $x->color === 'BLACK') {
            if ($x === $x->parent->left) {
                $w = $x->parent->right;
                if ($w->color === 'RED') {
                    $w->color = 'BLACK';
                    $x->parent->color = 'RED';
                    $this->leftRotate($x->parent);
                    $w = $x->parent->right;
                }
                if ($w->left->color === 'BLACK' && $w->right->color === 'BLACK') {
                    $w->color = 'RED';
                    $x = $x->parent;
                } else {
                    if ($w->right->color === 'BLACK') {
                        $w->left->color = 'BLACK';
                        $w->color = 'RED';
                        $this->rightRotate($w);
                        $w = $x->parent->right;
                    }
                    $w->color = $x->parent->color;
                    $x->parent->color = 'BLACK';
                    $w->right->color = 'BLACK';
                    $this->leftRotate($x->parent);
                    $x = $this->root;
                }
            } else {
                $w = $x->parent->left;
                if ($w->color === 'RED') {
                    $w->color = 'BLACK';
                    $x->parent->color = 'RED';
                    $this->rightRotate($x->parent);
                    $w = $x->parent->left;
                }
                if ($w->right->color === 'BLACK' && $w->left->color === 'BLACK') {
                    $w->color = 'RED';
                    $x = $x->parent;
                } else {
                    if ($w->left->color === 'BLACK') {
                        $w->right->color = 'BLACK';
                        $w->color = 'RED';
                        $this->leftRotate($w);
                        $w = $x->parent->left;
                    }
                    $w->color = $x->parent->color;
                    $x->parent->color = 'BLACK';
                    $w->left->color = 'BLACK';
                    $this->rightRotate($x->parent);
                    $x = $this->root;
                }
            }
        }
        $x->color = 'BLACK';
    }

    private function search(RBNode $node, int $key): RBNode {
        if ($node === $this->NIL || $key === $node->key) {
            return $node;
        }
        if ($key < $node->key) {
            return $this->search($node->left, $key);
        }
        return $this->search($node->right, $key);
    }

    private function inorderTraversal(RBNode $node, array &$result): void {
        if ($node !== $this->NIL) {
            $this->inorderTraversal($node->left, $result);
            $result[] = $node->key;
            $this->inorderTraversal($node->right, $result);
        }
    }

    public function insert(int $key): void {
        $z = new RBNode($key);
        $z->left = $this->NIL;
        $z->right = $this->NIL;

        $y = null;
        $x = $this->root;

        while ($x !== $this->NIL) {
            $y = $x;
            if ($z->key < $x->key) {
                $x = $x->left;
            } else {
                $x = $x->right;
            }
        }

        $z->parent = $y;
        if ($y === null) {
            $this->root = $z;
        } else if ($z->key < $y->key) {
            $y->left = $z;
        } else {
            $y->right = $z;
        }

        $this->insertFixup($z);
    }

    public function delete(int $key): void {
        $z = $this->search($this->root, $key);
        if ($z === $this->NIL) return;

        $y = $z;
        $yOriginalColor = $y->color;
        $x = null;

        if ($z->left === $this->NIL) {
            $x = $z->right;
            $this->transplant($z, $z->right);
        } else if ($z->right === $this->NIL) {
            $x = $z->left;
            $this->transplant($z, $z->left);
        } else {
            $y = $this->minimum($z->right);
            $yOriginalColor = $y->color;
            $x = $y->right;
            if ($y->parent === $z) {
                $x->parent = $y;
            } else {
                $this->transplant($y, $y->right);
                $y->right = $z->right;
                $y->right->parent = $y;
            }
            $this->transplant($z, $y);
            $y->left = $z->left;
            $y->left->parent = $y;
            $y->color = $z->color;
        }

        if ($yOriginalColor === 'BLACK') {
            $this->deleteFixup($x);
        }
    }

    public function searchKey(int $key): bool {
        return $this->search($this->root, $key) !== $this->NIL;
    }

    public function inorder(): array {
        $result = [];
        $this->inorderTraversal($this->root, $result);
        return $result;
    }
}

?>